extensions-contrib/kafka-emitter/src/main/java/org/apache/druid/emitter/kafka/KafkaEmitter.java

```diff
@@ -113,17 +121,42 @@ protected Producer<String, String> setKafkaProducer()
     }
   }
 
+  @VisibleForTesting
+  protected Producer<String, DruidSegmentEvent> setKafkaMetadataProducer()
+  {
+    ClassLoader currCtxCl = Thread.currentThread().getContextClassLoader();
+    try {
+      Thread.currentThread().setContextClassLoader(getClass().getClassLoader());
+
+      Properties props = new Properties();
+      props.put(ProducerConfig.BOOTSTRAP_SERVERS_CONFIG, config.getBootstrapServers());
+      props.put(ProducerConfig.KEY_SERIALIZER_CLASS_CONFIG, StringSerializer.class.getName());
+      props.put(ProducerConfig.VALUE_SERIALIZER_CLASS_CONFIG, DruidSegmentEventSerializer.class.getName());
+      props.put(ProducerConfig.RETRIES_CONFIG, DEFAULT_RETRIES);
+      props.putAll(config.getKafkaProducerConfig());
```

Comment:
I'm not convinced we need a completely separate producer for this, since it would unnecessarily double the number of connections to Kafka and make this harder to debug. The only reason we seem to do that is to pass in a different serializer, which can be addressed by moving the serialization concerns into the code itself and using the `ByteArraySerializer` for the value instead.

Reply:
the other benefit of using a single producer is that we can still give users the option to get metadata events in json format, similar to existing events if they want to. While we might like the proto format, it's unlikely this change would be accepted upstream, since it would be odd for metadata events to be sent in a different format by default.

Reply:
>which can be addressed by moving the serialization concerns into the code itself and using the ByteArraySerializer for the value instead.

If we change the existing behavior of using `StringSerializer` to `ByteArraySerializer` for non segment-metadata events(alerts, metrics and requests), won't it break the existing consumers of these Kafka topics?

Reply:
@harinirajendran it shouldn't, we would just move the serialization into the code itself to convert json strings to bytes and choose which format to use based on topic and configuration rather than relying on a static producer config. This means the Kafka producer would no longer be involved in the serialization.

It would also make the serialization a lot more efficient, since today we first serialize the json to string here https://github.com/apache/druid/blob/0bc9f9f30395309995a851fb4ed60a0521381b82/extensions-contrib/kafka-emitter/src/main/java/org/apache/druid/emitter/kafka/KafkaEmitter.java#L180
and then we convert that string to UTF-8 twice, once in the emitter https://github.com/apache/druid/blob/0bc9f9f30395309995a851fb4ed60a0521381b82/extensions-contrib/kafka-emitter/src/main/java/org/apache/druid/emitter/kafka/KafkaEmitter.java#L184, and another time as part of the StringSerializers. That's pretty inefficient.

By moving the serialization to bytes into the emitter directly, we can avoid that and have the jsonMapper write a byte array directly using writeValueAsBytes.

For topics using protos, we would switch to using the proto functions to convert to bytes directly as well.

Reply:
Thanks for the explanation Xavier. This makes sense. I have updated the code